### ISSUE TITLE
Vagrant Share stuck

### ISSUE THREAD

**Issue by paul-gene** [hashicorp/vagrant#10799]
### Vagrant version
2.2.4

### Host operating system
Mac OS 10.13.6

### Guest operating system
Debian Jessie

### Vagrantfile
```ruby
VAGRANTFILE_API_VERSION = "2"

Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
    config.vm.define "demo" do |web|
        web.vm.box = "debian/contrib-jessie64"
        web.vm.network :private_network, ip: "192.168.100.100"
        web.vm.synced_folder ".", "/var/www", type: "nfs", mount_options: ['rw', 'vers=3', 'tcp'], linux__nfs_options: ['rw','no_subtree_check','all_squash','async']
        web.vm.provision :shell, :path => "install.sh"
        web.vm.provider "virtualbox" do |v|
            v.memory = 4096
            v.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
            v.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
        end
    end
end
```

### Expected behavior
I should have been able to access the Vagrant VM via Ngrok

### Actual behavior
The Share Session was not started fully

### Steps to reproduce
1. Run `vagrant up`
2. Run `vagrant share --http 80`
3. Notice that the command gets stuck with the following output

```
Vagrant Share now defaults to using the `ngrok` driver.
The `classic` driver has been deprecated.

For more information about the `ngrok` driver, please
refer to the documentation:

  https://www.vagrantup.com/docs/share/

==> demo: Detecting network information for machine...
    demo: Local machine address: 127.0.0.1
    demo:
    demo: Note: With the local address (127.0.0.1), Vagrant Share can only
    demo: share any ports you have forwarded. Assign an IP or address to your
    demo: machine to expose all TCP ports. Consult the documentation
    demo: for your provider ('virtualbox') for more information.
    demo:
    demo: Local HTTP port: 80
    demo: Local HTTPS port: disabled
    demo: Port: 2222
==> demo: Creating Vagrant Share session...
```

**Comment by briancain** [comment]
Hi there,

Thank you for opening an issue. However, there is not enough information here to reproduce or provide a recommendation. When you first open an issue, a template is pre-filled in the box. That template provides instructions for getting your host and guest information, the version of Vagrant, and the debug output. Without that information, it is very challenging for us to assist you.

Please supply that information either by updating this ticket or by closing this ticket and opening a new one. Thank you!

**Comment by paul-gene** [comment]
Issue updated.

**Comment by briancain** [comment]
@paul-gene - Do you have the full debug log of running `vagrant share --http 80`?

**Comment by vitorrd** [comment]
I have the very same problem, but I have an extra question. When there's a local port exposed by Vagrant (through :forwarded_port), why doesn't Vagrant use it? It seems it always defaults to 80, even if a port was specifically forwarded through Vagrantfile.

On topic, by looking at the debug logs, I can see ngrok outputting an URL which does work, but Vagrant never outputs it as it seems some errors follow, but are not detrimental to the share working at all. Here are my logs on this, if it helps (I'm forcing the correct port as per the above paragraph):

`==> default: Creating Vagrant Share session...
 INFO subprocess: Starting process: ["D:\\Program Files\\Oracle\\VirtualBox\\VBoxManage.exe", "showvminfo", "5d5d3104-2dad-4e21-9c1f-8da72fbacddd", "--machinereadable"]
 INFO subprocess: Command not in installer, restoring original environment...
DEBUG subprocess: Selecting on IO
DEBUG subprocess: stdout: name="vagrant_getting_started_default_1555938275253_68561"
groups="/"
ostype="Ubuntu (64-bit)"
UUID="5d5d3104-2dad-4e21-9c1f-8da72fbacddd"
CfgFile="C:\\Users\\videl\\VirtualBox VMs\\vagrant_getting_started_default_1555938275253_68561\\vagrant_getting_started_default_1555938275253_68561.vbox"
SnapFldr="C:\\Users\\videl\\VirtualBox VMs\\vagrant_getting_started_default_1555938275253_68561\\Snapshots"
LogFldr="C:\\Users\\videl\\VirtualBox VMs\\vagrant_getting_started_default_1555938275253_68561\\Logs"
hardwareuuid="5d5d3104-2dad-4e21-9c1f-8da72fbacddd"
memory=384
pagefusion="off"
vram=8
cpuexecutioncap=100
hpet="off"
cpu-profile="host"
chipset="piix3"
firmware="BIOS"
cpus=2
pae="on"
longmode="on"
triplefaultreset="off"
apic="on"
x2apic="off"
nested-hw-virt="off"
cpuid-portability-level=0
bootmenu="messageandmenu"
boot1="disk"
boot2="dvd"
boot3="none"
boot4="none"
acpi="on"
ioapic="on"
biosapic="apic"
biossystemtimeoffset=0
rtcuseutc="on"
hwvirtex="on"
nestedpaging="on"
largepages="on"
vtxvpid="on"
vtxux="on"
paravirtprovider="legacy"
effparavirtprovider="none"
VMState="running"
VMStateChangeTime="2019-04-22T13:04:39.186000000"
monitorcount=1
accelerate3d="off"
accelerate2dvideo="off"
teleporterenabled="off"
teleporterport=0
teleporteraddress=""
teleporterpassword=""
tracing-enabled="off"
tracing-allow-vm-access="off"
tracing-config=""
autostart-enabled="off"
autostart-delay=0
defaultfrontend=""
storagecontrollername0="IDE Controller"
storagecontrollertype0="PIIX4"
storagecontrollerinstance0="0"
storagecontrollermaxportcount0="2"
storagecontrollerportcount0="2"
storagecontrollerbootable0="on"
storagecontrollername1="SATA Controller"
storagecontrollertype1="IntelAhci"
storagecontrollerinstance1="0"
storagecontrollermaxportcount1="30"
storagecontrollerportcount1="1"
storagecontrollerbootable1="on"
"IDE Controller-0-0"="emptydrive"
"IDE Controller-IsEjected"="off"
"IDE Controller-0-1"="none"
"IDE Controller-1-0"="emptydrive"
"IDE Controller-IsEjected"="off"
"IDE Controller-1-1"="none"
"SATA Controller-0-0"="C:\Users\videl\VirtualBox VMs\vagrant_getting_started_default_1555938275253_68561\box-disk1.vmdk"
"SATA Controller-ImageUUID-0-0"="c7334b32-c2ae-440a-ae75-1d5aeda49f7b"
natnet1="nat"
macaddress1="080027880CA6"
cableconnected1="on"
nic1="nat"
nictype1="82540EM"
nicspeed1="0"
mtu="0"
sockSnd="64"
sockRcv="64"
tcpWndSnd="64"
tcpWndRcv="64"
Forwarding(0)="ssh,tcp,127.0.0.1,2222,,22"
Forwarding(1)="tcp4567,tcp,,4567,,80"
nic2="none"
nic3="none"
nic4="none"
nic5="none"
nic6="none"
nic7="none"
nic8="none"
hidpointing="ps2mouse"
hidkeyboard="ps2kbd"
uart1="off"
uart2="off"
uart3="off"
uart4="off"
lpt1="off"
lpt2="off"
audio="none"
audio_in="on"
audio_out="on"
clipboard="disabled"
draganddrop="disabled"
SessionName="headless"
VideoMode="640,480,32"@0,0 1
vrde="off"
usb="off"
ehci="off"
xhci="off"
SharedFolderNameMachineMapping1="vagrant"
SharedFolderPathMachineMapping1="\\\\?\\C:\\Users\\videl\\Desktop\\vagrant_getting_started"
VRDEActiveConnection="off"
VRDEClients==0
videocap="off"
videocapaudio="off"
capturescreens="0"
capturefilename="C:\\Users\\videl\\VirtualBox VMs\\precise64_1555938271137_70393\\precise64_1555938271137_70393.webm"
captureres="1024x768"
capturevideorate=512
capturevideofps=25
captureopts=""
GuestMemoryBalloon=0
GuestOSType="Linux26_64"
GuestAdditionsRunLevel=2
GuestAdditionsVersion="4.2.0 r80737"
GuestAdditionsFacility_VirtualBox Base Driver=50,1555938284338
GuestAdditionsFacility_VirtualBox System Service=50,1555938287468
GuestAdditionsFacility_Seamless Mode=0,1555938284338
GuestAdditionsFacility_Graphics Mode=0,1555938284338
DEBUG subprocess: Waiting for process to exit. Remaining to timeout: 32000
DEBUG subprocess: Exit status: 0
DEBUG share: Generated configuration for ngrok:
---
tunnels:
  http:
    proto: http
    bind_tls: false
    addr: '4567'

DEBUG share: Starting ngrok proxy process.
 INFO subprocess: Starting process: ["D:\\bin/ngrok.EXE", "start", "--config", "C:/Users/videl/.ngrok2/ngrok.yml", "--config", "C:/Users/videl/AppData/Local/Temp/vagrant-share20190422-11896-1i6iy8y", "--all", "--log",
"stdout", "--log-format", "json", "--log-level", "debug"]
 INFO subprocess: Command not in installer, restoring original environment...
DEBUG subprocess: Selecting on IO
DEBUG subprocess: stdout: {"err":"\u003cnil\u003e","lvl":"info","msg":"open config file","path":"C:\\Users\\videl\\.ngrok2\\ngrok.yml","t":"2019-04-22T10:34:08.9254245-03:00"}
{"err":"\u003cnil\u003e","lvl":"info","msg":"open config file","path":"C:\\Users\\videl\\AppData\\Local\\Temp\\vagrant-share20190422-11896-1i6iy8y","t":"2019-04-22T10:34:08.9254245-03:00"}
{"lvl":"dbug","msg":"starting component","name":"signal handler","obj":"app","t":"2019-04-22T10:34:08.9264002-03:00"}
{"lvl":"dbug","msg":"starting component","name":"Tunnel session","obj":"app","t":"2019-04-22T10:34:08.9264002-03:00"}
{"lvl":"dbug","msg":"starting component","name":"web","obj":"app","t":"2019-04-22T10:34:08.9264002-03:00"}
{"lvl":"dbug","msg":"starting component","name":"updater","obj":"app","t":"2019-04-22T10:34:08.9264002-03:00"}
{"interval":"6h0m0s","lvl":"dbug","msg":"checking for updates periodically","obj":"updater","t":"2019-04-22T10:34:08.9264002-03:00"}
{"lvl":"dbug","msg":"check for update","obj":"updater","t":"2019-04-22T10:34:08.9264002-03:00"}
{"addr":"127.0.0.1:4040","lvl":"info","msg":"starting web service","obj":"web","t":"2019-04-22T10:34:08.9264002-03:00"}
{"err":"\u003cnil\u003e","id":"aaf0ae657fab","lvl":"dbug","msg":"open stream","obj":"csess","reqtype":0,"t":"2019-04-22T10:34:09.4289609-03:00"}
{"err":"\u003cnil\u003e","id":"aaf0ae657fab","lvl":"dbug","msg":"encode request","obj":"csess","req":{"Version":["2"],"ClientId":"","Extra":{"OS":"windows","Arch":"amd64","Authtoken":"3ufbxGH4tNW4StJK47UAS_7goK2zwv7gfg4HQ5AAN7q","Version":"2.3.25","Hostname":"tunnel.us.ngrok.com","UserAgent":"ngrok/2","Metadata":"","Cookie":"","HeartbeatInterval":10000000000,"HeartbeatTolerance":15000000000,"Fingerprint":null}},"sid":3,"t":"2019-04-22T10:34:09.4289609-03:00"}
{"err":"\u003cnil\u003e","id":"aaf0ae657fab","lvl":"dbug","msg":"decoded response","obj":"csess","resp":{"Version":"2","ClientId":"161a2cada06cd634a6625f03d27037b6","Error":"","Extra":{"Version":"prod","Cookie":"z/+JDvvqWzggrdD57mEGvSHcAn3QXvKc$DzKpRYme9UqVBWmgVpIvg6VZAQJE2Udmi3vfXxsn9DXlO0PtLScb1qi6nAZpm9iudxIqnOcCKWW22H/mv6LlG9FbU9VEk4QDYPfNU3Xa+c54ZCE7v4c3oD4lFT8RzGLy3i8PSZCoX5fBK6jtxlXWtIDY+hffVDyZO4dY/LbjBxBA0kqcGB1Ns8rStBiEBuMlTqUxM4socNJ0NNeV","AccountName":"vitorrd","SessionDuration":0,"PlanName":"Free"}},"sid":3,"t":"2019-04-22T10:34:09.5975316-03:00"}
{"lvl":"info","msg":"tunnel session started","obj":"tunnels.session","t":"2019-04-22T10:34:09.5975316-03:00"}
{"id":"2d7e5e69eb60","lvl":"info","msg":"client session established","obj":"csess","t":"2019-04-22T10:34:09.5975316-03:00"}
{"clientid":"161a2cada06cd634a6625f03d27037b6","err":"\u003cnil\u003e","id":"aaf0ae657fab","lvl":"dbug","msg":"open stream","obj":"csess","reqtype":1,"t":"2019-04-22T10:34:09.5975316-03:00"}
{"clientid":"161a2cada06cd634a6625f03d27037b6","err":"\u003cnil\u003e","id":"aaf0ae657fab","lvl":"dbug","msg":"encode request","obj":"csess","req":{"Id":"","Proto":"http","Opts":{"Hostname":"","Auth":"","Subdomain":"","HostHeaderRewrite":false,"LocalURLScheme":"http"},"Extra":{"Token":""}},"sid":7,"t":"2019-04-22T10:34:09.5975316-03:00"}
{"lvl":"dbug","msg":"no update available","obj":"updater","t":"2019-04-22T10:34:09.6239009-03:00"}
{"clientid":"161a2cada06cd634a6625f03d27037b6","err":"\u003cnil\u003e","id":"aaf0ae657fab","lvl":"dbug","msg":"decoded response","obj":"csess","resp":{"Id":"bbf74bb406555578f2eb3297e83b06cb","URL":"http://bf0c713d.ngrok.io","Proto":"http","Opts":{"Auth":"","HostHeaderRewrite":false,"Hostname":"bf0c713d.ngrok.io","LocalURLScheme":"http","Subdomain":""},"Error":"","Extra":{"Token":"jsJNJMoGni55U4CpH4pWtrzRcaO+slbS$hl2S7OqcaGvjuoYHmUpo7G6m95uAjM1iqZQavHY5E//OkPRS4BZUEqqZtlDjtaROx95d4wKORP6eCpihmtmyKRsPJwnPLdUzdTafKuNhmbpJkA7mp37aFMWsuMclRvR4B/gSqdSR/lCzFNl6vrbFxtKF0AhjqowePBV3XRwJLMdV4386bHvykLcX84a4VqfaiAxh3wrk8dZa4OQNFk0q7FWQFp2P+Skmb3AbztCffx17FvRGBc81LFOsNmFCDKdFqisTUtb2CS7m4+F4"}},"sid":7,"t":"2019-04-22T10:34:09.7683361-03:00"}
{"err":"\u003cnil\u003e","lvl":"dbug","msg":"start tunnel listen","name":"http","obj":"tunnels.session","opts":{"Hostname":"","Auth":"","Subdomain":"","HostHeaderRewrite":false,"LocalURLScheme":"http"},"proto":"http","t":"2019-04-22T10:34:09.7683361-03:00"}
{"addr":"http://localhost:4567","lvl":"info","msg":"started tunnel","name":"http","obj":"tunnels","t":"2019-04-22T10:34:09.7683361-03:00","url":"http://bf0c713d.ngrok.io"}
DEBUG subprocess: stdout: {"clientid":"161a2cada06cd634a6625f03d27037b6","id":"aaf0ae657fab","latency_ms":164,"lvl":"dbug","msg":"heartbeat received","obj":"csess","t":"2019-04-22T10:34:19.5941734-03:00"}`

I emphasize the ngrok URL does work, it's just never output by vagrant share.

Thanks and apologies for hijacking the issue.

**Comment by paul-gene** [comment]
@briancain how can I get a debug log when I run the command?

**Comment by paul-gene** [comment]
I can see the Ngrok URL in the debug too, but when I visit it, I get the message:

> Failed to complete tunnel connection
> The connection to http://08f789a1.ngrok.io was successfully tunneled to your ngrok client, but the client failed to establish a connection to the local address localhost:80.
> 
> Make sure that a web service is running on localhost:80 and that it is a valid address.
> 
> The error encountered was: dial tcp [::1]:80: connect: connection refused

How do I run `vagrant share` pointing towards my vhost I have setup for the Vagrant Box?

**Comment by briancain** [comment]
@paul-gene - hmmm...is this config file empty at all? Or does it exist?

`/Users/paulcanning/.ngrok2/ngrok.yml`

from...

> DEBUG subprocess: stdout: {"err":"\u003cnil\u003e","lvl":"info","msg":"open config file","path":"/Users/paulcanning/.ngrok2/ngrok.yml","t":"2019-04-26T10:48:07.115148+01:00"}

**Comment by paul-gene** [comment]
@briancain it exists, but simply has an auth token in it

**Comment by paul-gene** [comment]
To add; when debugging, the ngrok URL is visible in the output. If I enter the URL in my phone, it manages to "load" the URL, but ends up redirecting to the vhost domain

eg

VHost = mydomain.test
Ngrok = 123456.ngrok.io

Enter 123456.ngrok.io into phone browser, blank page loads slowly then ends up on mydomain.test

**Comment by briancain** [comment]
Thanks for the additional information! I'll put it in our next milestone to be worked on for the next release. :heart:

**Comment by oeramirez** [comment]
Hi. I confirm this bug. I was following the Vagrant Getting Started tutorial and it took me a while to find the ngrok URL buried in the debug log:

{"addr":"http://localhost:4567","lvl":"info","msg":"started tunnel","name":"http","obj":"tunnels","t":"2019-07-13T16:00:20.9730045+02:00","url":"http://44852af9.ngrok.io"}

It is not shown however in the normal output, leading to believe that the share is not working:

> vagrant share
> Vagrant Share now defaults to using the `ngrok` driver.
> The `classic` driver has been deprecated.
> 
> For more information about the `ngrok` driver, please
> refer to the documentation:
> 
>   https://www.vagrantup.com/docs/share/
> ==> default: Detecting network information for machine...
>     default: Local machine address: 127.0.0.1
>     default:
>     default: Note: With the local address (127.0.0.1), Vagrant Share can only
>     default: share any ports you have forwarded. Assign an IP or address to your
>     default: machine to expose all TCP ports. Consult the documentation
>     default: for your provider ('virtualbox') for more information.
>     default:
>     default: Local HTTP port: 4567
>     default: Local HTTPS port: disabled
>     default: Port: 2222
>     default: Port: 4567
> ==> default: Creating Vagrant Share session...

**Comment by yangshuairocks** [comment]
I have the same issue just as @oeramirez described. I'm on Windows 10 with vagrant 2.2.5.

**Comment by jtinsky** [comment]
+1 on this bug with Vagrant version 2.2.4 and ngrok version 2.3.34 on MacOS 10.14.6

**Comment by fqr297** [comment]
+1 on this bug with Vagrant version 2.2.5 and ngrok version 2.3.34 on MacOS 10.14.6

**Comment by jacklenox** [comment]
Another +1 on this bug with Vagrant version 2.2.5, and ngrok version 2.3.34 on Ubuntu 18.04. As per @fqr297's suggestion, I was able to find the ngrok URL in debug mode.

**Comment by pe5er** [comment]
+1 on this as well; same as above

**Comment by sameerast** [comment]
+1 on this as well; same as above

**Comment by vrubiolo** [comment]
I am seeing the same issue here with Vagrant 2.2.6 on Fedora 30 with a newly created ngrok account. As for other reporters, I can get the ngrok URL by using `--debug` but it is not shown otherwise.

Looking at https://github.com/Chassis/Chassis/issues/644, this happens to be caused by a bug/change in `ngrok` itself, which was introduced in versions after 2.2.8. I can confirm that using ngrok v2.2.8 (using the link in the abovementioned Chassis issue) allows the ngrok URL to be displayed with `vagrant share`

Maybe a mention in the [Vagrant share docs](https://www.vagrantup.com/docs/share/) could be useful too, just as they did for Chassis?

**Comment by FerRubioMorales** [comment]
Hmmm some fix? I upgraded to the latest vagran and i'ts only work with --debug

**Comment by meitar** [comment]
I have the same issue with Vagrant version 2.2.7 and ngrok version 2.3.35. At least, the symptoms are the same. In my case, my [ngrok dashboard status screen](https://dashboard.ngrok.com/status) shows that I have a tunnel, but when I try actually loading the URL shown there, the ngrok service tells me that "Tunnel TUNNEL_ID.ngrok.io not found." On the Vagrant side, Vagrant's output hangs at "`==> default: Creating Vagrant Share session...`" as it does for the others in this thread.

For what it's worth, you can see more of ngrok's data by loading http://localhost:4040 after starting the `vagrant share` attempt, even if Vagrant fails to complete the connection.

**Comment by m-faheem** [comment]
I am having the same issue with Vagrant 2.2.7, ngrok version 2.3.35 and macOS 10.15.3, any fix yet!

**Comment by benbowler** [comment]
This integration seems to be deprecated, you can now just use ngrok directly. `ngrok http 197.168.10.10:8000` where the IP and port are those of you vagrant box, you can also use any local domains you have mapped in your hosts file. Amazing system!

**Comment by shannoncarver** [comment]
I'm also having the issue on Vagrant 2.2.7 and macOS 10.15.4

**Comment by chrisroberts** [comment]
This has been fixed in vagrant-share 1.1.10

Cheers!
